### PR TITLE
Database Automatic User Provisioning support for self-hosted MongoDB

### PR DESCRIPTION
**Pull request by greedy52**
Part of #27323 

changelog: Database Automatic User Provisioning support for self-hosted MongoDB

Implemented as RFD: #33750

Testing:

- [x] Modes
  - [x] keep
  - [x] drop
- [x] Single self-hosted MongoDB
  - [x] Mongo 7.0
  - [x] Mongo 6.0
  - [x] Mongo 5.0
  - [x] Mongo 4.4
  - [x] Mongo 4.2
  - [x] Mongo 4.0 

<details>
<summary>Manual testing example </summary>

#### 1. Configure self-hosted MongoDB

Setup a self-hosted MongoDB. Sample docker setup
https://github.com/greedy52/teleport-database-test-setup/tree/main/mongo

Log into the database as the default admin, for ex: 
```
docker exec -it mongo mongosh -u mongoadmin -p secret
```

Create the role for `teleport-admin`:
```
db.getSiblingDB("admin").runCommand({
    createRole: "teleport-admin-role",
    privileges: [
        { resource: { cluster: true }, actions: [ "inprog" ] },
        { resource: { db: "", collection: "" }, actions: [ "grantRole", "revokeRole" ] },
        { resource: { db: "$external", "collection": "" }, actions: [ "createUser", "updateUser", "dropUser", "viewUser", "setAuthenticationRestriction", "changeCustomData"] },
    ],
    roles: [],
})
```

Create `teleport-admin` user:
```
db.getSiblingDB("$external").runCommand({
  createUser: "CN=teleport-admin",
  roles: [ { role: 'teleport-admin-role', db: 'admin' } ],
})
```

Create a custom role for testing:
```
db.getSiblingDB("test").createRole({role: "my-custom-role", privileges:[], roles:[]});
```

#### 2. Configure Teleport
Create a Teleport role for auto-user and assign it to your Teleport user, ex:
```yaml
kind: role
version: v6
metadata:
  name: mongo-auto-user
spec:
  options:
    create_db_user_mode: keep
  allow:
    db_labels:
      automongo: "true"
    db_names:
    - "*"
    db_roles:
    - read@test2
    - readWrite@test
    - my-custom-role@test
```
Assign the role to your User.

Create a database:
```
  - name: "self-hosted-mongo-auto"
    protocol: "mongodb"
    uri: "localhost:27017"
    static_labels:
      automongo: "true"
    admin_user:
      name: "teleport-admin"
```

#### 3. Connect
- `tsh login`
- `tsh db connect --db-user <teleport-user> --db-name test self-hosted-mongo-auto`
```
test> db.users.insertOne({name:"steve", age: 99})
{
  acknowledged: true,
  insertedId: ObjectId("...")
}
test> use $external
$external> db.getUser("CN=<teleport-user>",{showCustomData:1})
{
  _id: '$external.CN=<teleport-user>',
  userId: new UUID("..."),
  user: 'CN=STeve',
  db: '$external',
  customData: { 'teleport-auto-user': true },
  roles: [
    { role: 'readWrite', db: 'test' },
    { role: 'read', db: 'test2' },
    { role: 'my-custom-role', db: 'test' }
  ],
  mechanisms: [ 'external' ]
}
```
</details>

The change is ready but waiting for #34132 to merge before open for review.
Also need #34271 to show proper error messages when setup is bad.

### CODE REVIEW

File: lib/srv/db/mongodb/autousers_admin.go [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+	})
+
+	client, err := mongo.Connect(ctx, clientCfg)
+	return newBasicAdminClient(client), trace.Wrap(err)
+}
+
+// basicAdminClient is a simple wrapper of mongo.Client with some helper
+// functions.
+type basicAdminClient struct {
+	*mongo.Client
```
Comment (smallinsky) [inline lib/srv/db/mongodb/autousers_admin.go:201]: At first I was afraid that mongo.Client is not thread safe but found this comment in mongo library:

```
// Client is a handle representing a pool of connections to a MongoDB deployment. It is safe for concurrent use by
// multiple goroutines.

type Client struct {
```

File: lib/srv/db/mongodb/autousers_admin.go [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+
+	shareableClient, err := getShareableAdminClient(ctx, adminClientFnCache, sessionCtx, e, makeBasicAdminClient)
+	return shareableClient, trace.Wrap(err)
+}
+
+func getShareableAdminClient(ctx context.Context, cache *utils.FnCache, sessionCtx *common.Session, e *Engine, makeBasicClient makeBasicAdminClientFunc) (*shareableAdminClient, error) {
+	key := struct {
+		databaseName string
+		databaseURI  string
+		adminUser    string
```
Comment (smallinsky) [inline lib/srv/db/mongodb/autousers_admin.go:64]: Probably there is a corner case during CA rotation where a shared client should be reused but a connection needs to be recreated but this should be really narrow case.

File: lib/srv/db/mongodb/autousers.go [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+type userRole struct {
+	Rolename string `bson:"role"`
+	Database string `bson:"db"`
+}
+
+type userRoles []userRole
+
+func (r userRoles) sort() {
+	slices.SortFunc(r, func(a, b userRole) int {
+		if cmpDatabase := cmp.Compare(a.Database, b.Database); cmpDatabase != 0 {
```
Comment (smallinsky) [inline lib/srv/db/mongodb/autousers.go:66]: nit:  Does string comparison < won't work in this case ?

File: lib/srv/db/mongodb/autousers.go [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+	for _, authRestriction := range u.AuthRestrictions {
+		if slices.Contains(authRestriction.ClientSource, lockedClientSource) {
+			return true
+		}
+	}
+	return false
+}
+
+type userCustomData struct {
+	TeleportAutoUser bool `bson:"teleport-auto-user"`
```
Comment (smallinsky) [inline lib/srv/db/mongodb/autousers.go:50]: nit: go docs

File: lib/srv/db/mongodb/autousers_admin.go [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+	c.log.Debugf("Created new MongoDB connection as admin user %q on %q.", c.adminUser, c.databaseName)
+
+	// Wait until TTL.
+	<-c.timer.Chan()
+
+	// Wait until all shared sessions are released.
+	c.wg.Wait()
+
+	// Disconnect.
+	if err := c.adminClient.Disconnect(context.Background()); err != nil {
```
Comment (smallinsky) [inline lib/srv/db/mongodb/autousers_admin.go:157]: nit: Can we update the comment for a reader: 
```suggestion
	// Disconnect connection. This happens after cache item expired to ensure that shared client won't be reused when wrapped client was disconnected.  
	if err := c.adminClient.Disconnect(context.Background()); err != nil {
```

**Review by smallinsky — APPROVED** [review]
Great job. Tested on my local setup without any issue. LGTM

File: lib/srv/db/mongodb/autousers_admin.go (hunk 2 above)
Comment (greedy52) [reply lib/srv/db/mongodb/autousers_admin.go:64]: Good catch. During a rotation, the server should be restarted with a new configuration so the connection will be dead. The connection could be killed for other reasons as well.

Let me try to add a check on whether the shared client is dead after getting it from the cache. If so, just make a non-shared client.

File: lib/srv/db/mongodb/autousers_admin.go (hunk 1 above)
Comment (greedy52) [reply lib/srv/db/mongodb/autousers_admin.go:201]: Yes. Let me add comment on this.

File: lib/srv/db/mongodb/autousers_admin.go (hunk 2 above)
Comment (greedy52) [reply lib/srv/db/mongodb/autousers_admin.go:64]: Seems the mongo client connection pool has the ability to reconnect even if i restarted mongo docker. So I will not handle that case.

As for the CA rotation case, users should understand that the connection may have issues during rotation. And it should self-heal after cache expires (one minute). I will leave a comment in the code for this case but not worth it to fix properly.

File: lib/srv/db/mongodb/autousers.go (hunk 3 above)
Comment (greedy52) [reply lib/srv/db/mongodb/autousers.go:81]: the compare function expects an `int` return, so it's easier to use `cmp.Compare` in my opinion and use it for return in next line.

**Review by gabrielcorado — APPROVED** [review]
I've also tested with my local setup and everything worked as expected. Nice work.

**Pull request merged**